Race Condition: MQTT Channel Closure Can Cause Panic

# Race Condition: MQTT Channel Closure Can Cause Panic

## Summary

The MQTT channel's `Stop()` method closes `inbox` while MQTT message handlers (`handleMessage`, `handleStatus`) may still be running and attempting to send to it, causing a potential panic from sending on a closed channel.

## Location

**File:** `internal/channels/mqtt.go`  
**Method:** `MQTTChannel.Stop()`

## Bug Details

### Current Implementation

```go
func (m *MQTTChannel) Stop() error {
	m.logger.Info("stopping mqtt channel")

	if m.client != nil && m.client.IsConnected() {
		m.client.Disconnect(250)
	}

	if m.cancel != nil {
		m.cancel()
	}

	m.wg.Wait()      // ⚠️ WaitGroup is never populated (no wg.Add calls)
	close(m.inbox)   // ⚠️ Channel closed while handlers may still send
	return nil
}
```

### The Race Condition

1. **MQTT message arrives** → Paho library invokes `handleMessage` in a separate goroutine
2. **`Stop()` is called** concurrently:
   - Calls `m.client.Disconnect(250)` (async, 250ms grace period)
   - Calls `m.cancel()` to signal context cancellation
   - Calls `m.wg.Wait()` (no-op since WaitGroup is never used)
   - **Immediately closes `m.inbox`**
3. **Handler tries to send:**
   ```go
   case m.inbox <- msg:  // 💥 PANIC: send on closed channel
   ```

### Why This Happens

- `client.Disconnect(250)` is **asynchronous** and gives a 250ms grace period
- During this window, the Paho MQTT library may still deliver queued messages
- The message handlers (`handleMessage`, `handleStatus`) are callbacks running in goroutines managed by the Paho library
- `m.wg` is declared but **never incremented** (`wg.Add` is never called), so `wg.Wait()` returns immediately
- `close(m.inbox)` executes before handlers finish processing

## Reproduction

### Minimal Test Case

```go
func TestMQTTShutdownRace(t *testing.T) {
	// Start MQTT channel
	mqtt := NewMQTT("localhost", 1883, "", "", logger)
	mqtt.Start(context.Background())
	
	// Simulate message flood
	go func() {
		for i := 0; i < 100; i++ {
			// Publish message that triggers handleMessage
			time.Sleep(1 * time.Millisecond)
		}
	}()
	
	time.Sleep(10 * time.Millisecond)
	mqtt.Stop() // Race: handlers still processing, inbox gets closed
}
```

### Production Trigger

This will manifest in production when:
- The orchestrator shuts down (SIGTERM, SIGINT)
- MQTT messages are actively arriving
- Timing is such that `Stop()` closes the channel while handlers are sending

**Crash log:**
```
panic: send on closed channel
goroutine 47 [running]:
github.com/clawinfra/evoclaw/internal/channels.(*MQTTChannel).handleMessage(...)
	internal/channels/mqtt.go:XXX
```

## Fix Strategy

### Option 1: Synchronize Handler Lifecycle (Recommended)

Track active handlers with the WaitGroup:

```go
func (m *MQTTChannel) handleMessage(client mqtt.Client, mqttMsg mqtt.Message) {
	m.wg.Add(1)
	defer m.wg.Done()
	
	m.logger.Debug("mqtt message received", "topic", mqttMsg.Topic())
	
	// ... existing parsing logic ...
	
	select {
	case m.inbox <- msg:
		m.logger.Debug("message queued")
	case <-m.ctx.Done():
		return
	default:
		m.logger.Warn("inbox full, dropping message")
	}
}

func (m *MQTTChannel) handleStatus(client mqtt.Client, mqttMsg mqtt.Message) {
	m.wg.Add(1)
	defer m.wg.Done()
	
	// ... existing logic ...
}

func (m *MQTTChannel) Stop() error {
	m.logger.Info("stopping mqtt channel")

	// 1. Stop accepting new messages
	if m.cancel != nil {
		m.cancel()
	}

	// 2. Disconnect MQTT (async)
	if m.client != nil && m.client.IsConnected() {
		m.client.Disconnect(250)
	}

	// 3. Wait for in-flight handlers to finish
	m.wg.Wait()
	
	// 4. NOW safe to close
	close(m.inbox)
	
	return nil
}
```

### Option 2: Check Before Send (Defensive)

Add a closed flag:

```go
type MQTTChannel struct {
	// ... existing fields ...
	closed   atomic.Bool
}

func (m *MQTTChannel) handleMessage(client mqtt.Client, mqttMsg mqtt.Message) {
	if m.closed.Load() {
		return
	}
	
	// ... existing logic ...
}

func (m *MQTTChannel) Stop() error {
	m.closed.Store(true)
	// ... rest of Stop logic ...
}
```

**Recommendation:** Use **Option 1** for correctness + graceful shutdown. Option 2 is a quick patch but doesn't guarantee all messages are processed.

## Impact

- **Severity:** High (can crash the orchestrator)
- **Affected:** All MQTT deployments (edge agent communication)
- **Workaround:** None (crash is non-recoverable)

## Testing

After fix, add integration test:

```go
func TestMQTTGracefulShutdown(t *testing.T) {
	// Verify no panic when stopping under message load
	// Verify all in-flight messages are processed or cleanly dropped
}
```

---

**Related Code:**
- `internal/orchestrator/orchestrator.go:receiveFrom()` (consumer side)
- `edge-agent` MQTT client (sender side)

**Environment:**
- Go 1.24
- Paho MQTT v1.5.0 (async callback model)

--- a/cmd/evoclaw/gateway.go
+++ b/cmd/evoclaw/gateway.go
@@ -1,11 +1,8 @@
 package main
 
 import (
-	"context"
 	"fmt"
-	"log/slog"
 	"os"
-	"os/signal"
 	"path/filepath"
 	"syscall"
 	"time"

--- a/cmd/evoclaw/main.go
+++ b/cmd/evoclaw/main.go
@@ -532,26 +532,18 @@ func printBanner(app *App) {
 // waitForShutdown waits for termination signal and performs graceful shutdown
 func waitForShutdown(app *App) error {
 	sigCh := make(chan os.Signal, 1)
-	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGUSR1)
+	signal.Notify(sigCh, getShutdownSignals()...)
 
 	for {
 		sig := <-sigCh
 		
-		switch sig {
-		case syscall.SIGHUP:
-			app.Logger.Info("reload signal received - config reload not yet implemented")
-			// TODO: Reload config without restart
+		// Handle platform-specific signals (SIGHUP, SIGUSR1 on Unix)
+		if handlePlatformSignal(sig, app.Logger) {
 			continue
-			
-		case syscall.SIGUSR1:
-			app.Logger.Info("update signal received - self-update not yet implemented")
-			// TODO: Trigger self-update and restart
-			continue
-			
-		case syscall.SIGINT, syscall.SIGTERM:
-			app.Logger.Info("shutdown signal received", "signal", sig)
-			// Break out of loop to perform shutdown
 		}
+		
+		// SIGINT or SIGTERM - proceed to shutdown
+		app.Logger.Info("shutdown signal received", "signal", sig)
 		break
 	}
 

--- a/cmd/evoclaw/signals_unix.go
+++ b/cmd/evoclaw/signals_unix.go
@@ -10,6 +10,24 @@ import (
 	"syscall"
 )
 
+// getShutdownSignals returns the signals to listen for on Unix systems
+func getShutdownSignals() []os.Signal {
+	return []os.Signal{syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGUSR1}
+}
+
+// handlePlatformSignal handles platform-specific signals, returns true if should continue loop
+func handlePlatformSignal(sig os.Signal, logger *slog.Logger) bool {
+	switch sig {
+	case syscall.SIGHUP:
+		logger.Info("reload signal received - config reload not yet implemented")
+		return true // continue loop
+	case syscall.SIGUSR1:
+		logger.Info("update signal received - self-update not yet implemented")
+		return true // continue loop
+	}
+	return false // don't continue, proceed to shutdown
+}
+
 func setupSignalHandlers(ctx context.Context, cancel context.CancelFunc, logger *slog.Logger) {
 	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGUSR1)

--- a/cmd/evoclaw/signals_windows.go
+++ b/cmd/evoclaw/signals_windows.go
@@ -10,6 +10,17 @@ import (
 	"syscall"
 )
 
+// getShutdownSignals returns the signals to listen for on Windows
+func getShutdownSignals() []os.Signal {
+	return []os.Signal{syscall.SIGINT, syscall.SIGTERM}
+}
+
+// handlePlatformSignal handles platform-specific signals, returns true if should continue loop
+func handlePlatformSignal(sig os.Signal, logger *slog.Logger) bool {
+	// Windows only handles SIGINT and SIGTERM, no special cases
+	return false
+}
+
 func setupSignalHandlers(ctx context.Context, cancel context.CancelFunc, logger *slog.Logger) {
 	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

--- a/internal/channels/mqtt.go
+++ b/internal/channels/mqtt.go
@@ -123,12 +123,12 @@ func (m *MQTTChannel) Start(ctx context.Context) error {
 func (m *MQTTChannel) Stop() error {
 	m.logger.Info("stopping mqtt channel")
 
-	if m.client != nil && m.client.IsConnected() {
-		m.client.Disconnect(250)
-	}
-
 	if m.cancel != nil {
 		m.cancel()
+	}
+
+	if m.client != nil && m.client.IsConnected() {
+		m.client.Disconnect(250)
 	}
 
 	m.wg.Wait()
@@ -202,6 +202,9 @@ func (m *MQTTChannel) subscribe() error {
 
 // handleMessage processes incoming messages from agents
 func (m *MQTTChannel) handleMessage(client mqtt.Client, mqttMsg mqtt.Message) {
+	m.wg.Add(1)
+	defer m.wg.Done()
+
 	m.logger.Debug("mqtt message received", "topic", mqttMsg.Topic())
 
 	var payload struct {
@@ -245,6 +248,9 @@ func (m *MQTTChannel) handleMessage(client mqtt.Client, mqttMsg mqtt.Message) {
 
 // handleStatus processes agent heartbeat/status updates
 func (m *MQTTChannel) handleStatus(client mqtt.Client, mqttMsg mqtt.Message) {
+	m.wg.Add(1)
+	defer m.wg.Done()
+
 	m.logger.Debug("agent status update", "topic", mqttMsg.Topic())
 
 	var status struct {

